File: solutions/javascript/sieve/sieve.js
// @ts-check

/**
 * Sieve of Eratosthenes algorithm for finding all prime numbers up to the given limit.
 *
 * @param {Number} limit
 * @returns {Number[]} prime numbers.
 */
export const primes = (limit) => {
  if (limit < 2) {
    return [];
  }

  const multiples = new Set([0, 1]);
  for (let n = 2; n < Math.sqrt(limit); n += 1) {
    for (let m = n * n; m <= limit; m += n) {
      multiples.add(m);
    }
  }

  const primes = new Set();
  [...Array(limit + 1).keys()].forEach((number) => {
    if (!multiples.has(number)) {
      primes.add(number);
    }
  });

  return Array.from(primes);
};

File: solutions/python/sieve/sieve.py
"""Function to find prime numbers up to the given limit."""


def primes(limit):
    """Sieve of Eratosthenes algorithm for finding all prime numbers.

    :param limit: int - given limit to find primes less than.
    :return: list - sorted list of prime numbers.
    """
    if limit < 2:
        return []
    numbers = set(range(3, limit + 1, 2))
    multiples = {m for n in numbers for m in range(n * n, limit + 1, 2 * n)}
    return [2] + sorted(numbers - multiples)
